stability: raft entry log cache to avoid busy waiting fetching last t…

spencerkimball opened the pull request:
…erms

Resurrection of #3189 to avoid busy-looping fetching last terms when
a node comes up alone in what used to be a multi-node cluster. The
original PR used a cache per Raft group. This one uses a single cache
and evicts based on total bytes stored. This change also stores newly-
fetched entries in the cache instead of only when they're appended to
the log.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8494)

<!-- Reviewable:end -->

spencerkimball committed:
stability: raft entry log cache to avoid busy waiting fetching last terms

Resurrection of #3189 to avoid busy-looping fetching last terms when
a node comes up alone in what used to be a multi-node cluster. The
original PR used a cache per Raft group. This one uses a single cache
and evicts based on total bytes stored. This change also stores newly-
fetched entries in the cache instead of only when they're appended to
the log.

spencerkimball commented:
As previously discussed, we probably want to wait on this due to larger stability concerns.

spencerkimball commented:
```
name                                            old time/op  new time/op  delta
Insert1_Cockroach-4                              396µs ±11%   388µs ± 4%     ~       (p=0.606 n=9+8)
Insert1Multinode_Cockroach-4                    1.08ms ±14%  0.82ms ±11%  -23.97%  (p=0.000 n=10+10)
Insert10_Cockroach-4                             500µs ± 1%   424µs ± 2%  -15.31%    (p=0.000 n=9+8)
Insert10Multinode_Cockroach-4                   1.16ms ± 3%  0.91ms ± 1%  -21.54%    (p=0.000 n=8+8)
Insert100_Cockroach-4                           1.50ms ± 2%  1.33ms ± 2%  -11.45%    (p=0.000 n=9+8)
Insert100Multinode_Cockroach-4                  2.61ms ± 5%  2.27ms ±17%  -12.95%  (p=0.001 n=10+10)
Insert1000_Cockroach-4                          10.6ms ± 2%   9.5ms ± 3%   -9.80%   (p=0.000 n=10+9)
Insert1000Multinode_Cockroach-4                 16.1ms ± 7%  13.5ms ± 6%  -16.15%  (p=0.000 n=10+10)
Update10_Cockroach-4                            829µs ± 3%   839µs ± 2%     ~     (p=0.063 n=10+10)
Update10Multinode_Cockroach-4                   1.36ms ± 3%  1.27ms ± 2%   -6.34%   (p=0.000 n=10+9)
Update100_Cockroach-4                           3.54ms ± 2%  3.64ms ± 2%   +2.86%  (p=0.000 n=10+10)
Update100Multinode_Cockroach-4                  4.68ms ± 4%  4.42ms ± 3%   -5.47%   (p=0.000 n=8+10)
Update1000_Cockroach-4                          31.7ms ± 6%  27.3ms ± 3%  -13.88%  (p=0.000 n=10+10)
Update1000Multinode_Cockroach-4                 38.6ms ± 7%  32.4ms ± 8%  -16.04%  (p=0.000 n=10+10)
Upsert1_Cockroach-4                             1.86ms ± 5%  1.53ms ± 6%  -17.92%  (p=0.000 n=10+10)
Upsert10_Cockroach-4                            2.41ms ± 1%  2.03ms ± 4%  -15.71%    (p=0.000 n=8+9)
Upsert100_Cockroach-4                           7.07ms ± 7%  6.36ms ±13%  -10.10%   (p=0.006 n=9+10)
Upsert1000_Cockroach-4                          49.2ms ± 1%  45.1ms ± 4%   -8.40%    (p=0.000 n=8+9)
Upsert10Multinode_Cockroach-4                   5.26ms ± 5%  4.26ms ± 5%  -19.06%  (p=0.000 n=10+10)
Upsert100Multinode_Cockroach-4                  11.8ms ± 7%  10.1ms ± 7%  -14.31%   (p=0.000 n=8+10)
Upsert1000Multinode_Cockroach-4                 70.1ms ± 7%  61.1ms ±15%  -12.81%  (p=0.000 n=10+10)
Delete1_Cockroach-4                              807µs ± 5%   819µs ± 4%     ~     (p=0.143 n=10+10)
Delete1Multinode_Cockroach-4                    1.46ms ± 5%  1.34ms ± 5%   -8.55%  (p=0.000 n=10+10)
Delete10_Cockroach-4                             964µs ± 4%   931µs ± 3%   -3.45%  (p=0.001 n=10+10)
Delete10Multinode_Cockroach-4                   1.75ms ± 8%  1.49ms ± 5%  -15.03%  (p=0.000 n=10+10)
Delete100_Cockroach-4                           2.53ms ± 5%  2.34ms ± 3%   -7.81%   (p=0.000 n=10+9)
Delete100Multinode_Cockroach-4                  3.70ms ± 7%  3.12ms ± 6%  -15.54%  (p=0.000 n=10+10)
Delete1000_Cockroach-4                          16.9ms ± 6%  13.9ms ± 7%  -17.84%  (p=0.000 n=10+10)
Delete1000Multinode_Cockroach-4                 22.1ms ± 2%  20.0ms ± 5%   -9.50%   (p=0.000 n=8+10)
```

bdarnell commented:
The profile uploaded to #8544 shows that we're spending 15% of our CPU time on the beta cluster in these `Term()` calls, so this is a priority now.

bdarnell commented:
Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[storage/entry_cache.go, line 16 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPEGEk3ej5mdsO5U0UL:-KPEGEk3ej5mdsO5U0UM:bvk5u42) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/entry_cache.go#L16)):*
> ```Go
> // for names of contributors.
> //
> // Author: Kevin GuanJian (keynovo@gmail.com)
> ```

Add your name here.

---

*[storage/entry_cache.go, line 53 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPEGWO_yQqNQQZQ7XyL:-KPEGWO_yQqNQQZQ7XyM:bu9vs4c) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/entry_cache.go#L53)):*
> ```Go
> 	// We expect the raft entry cache mutex to be held on ShouldEvict
> 	// and OnEvicted callbacks.
> 	// On ShouldEvice, compare the total size of the cache in bytes to the
> ```

s/Evice/Evict/

---

*[storage/entry_cache.go, line 62 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPEJ6dvx_-Okv5urjSQ:-KPEJ6dvx_-Okv5urjSR:bglatxb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/entry_cache.go#L62)):*
> ```Go
> 	rec.cache.Config.OnEvicted = func(k, v interface{}) {
> 		ent := v.(raftpb.Entry)
> 		rec.bytes -= uint64(ent.Size())
> ```

We need to update `firstIndex` on eviction. 

---

*[storage/entry_cache.go, line 87 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPEGxmkKHOXG-hrwvkw:-KPEGxmkKHOXG-hrwvkx:b-taalgg) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/entry_cache.go#L87)):*
> ```Go
> }
> 
> // getEntries returns entries between [lo, hi) for specified range.
> ```

Comment that if any entries are returned, they will start with `lo` and proceed sequentially without gaps. 

---

*[storage/entry_cache.go, line 89 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPEHRYbsbrvvGe2E_CP:-KPEHRYbsbrvvGe2E_CQ:b-y50ais) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/entry_cache.go#L89)):*
> ```Go
> // getEntries returns entries between [lo, hi) for specified range.
> func (rec *raftEntryCache) getEntries(rangeID roachpb.RangeID, lo, hi, maxBytes uint64) (
> 	[]raftpb.Entry, uint64 /* size in bytes */, uint64 /* max Index */) {
> ```

s/max/next/ if I'm reading this correctly (or subtract one from all the return values here and add it on the caller side)

---

*[storage/entry_cache.go, line 124 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPEILMeKKconR1tkWlT:-KPEILMeKKconR1tkWlU:b-tyh9p2) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/entry_cache.go#L124)):*
> ```Go
> 			rec.cache.Del(entryCacheKey{RangeID: rangeID, Index: i})
> 		}
> 		rec.firstIndexes[rangeID] = hi
> ```

There's no guarantee that `hi` is in the cache (and with this data structure we have no way of finding the next remaining index). If there are multiple discontinuous cached spans, we'll have a `firstIndex` that is below the actual first index. This could make `clearTo` unexpectedly expensive as it has to iterate from an old `firstIndex` to the requested index. Maybe this should be an `OrderedCache` so we don't have to track the `firstIndexes` separate from the cache contents.

---

*[storage/replica.go, line 1636 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPED7UBqPJhXvzMGeqV:-KPED7UBqPJhXvzMGeqW:bj6x7j1) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/replica.go#L1636)):*
> ```Go
> 	// log entries and cache the latest ones.
> 	r.mu.Lock()
> 	r.store.raftEntryCache.delEntries(r.RangeID, lastIndex+1, r.mu.lastIndex)
> ```

It's not clear to me that this is beneficial. If the entries are not yet committed then they will probably be read again so it's worth leaving them in the cache. I'd guess that we want to delete entries up to the (new) applied index instead of the (prevous) last index.

---

*[storage/replica_raftstorage.go, line 200 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPEFCFogmqAEVo3i_zH:-KPEFCFogmqAEVo3i_zI:b-oz094r) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/replica_raftstorage.go#L200)):*
> ```Go
> // Term requires that the replica lock is held.
> func (r *Replica) Term(i uint64) (uint64, error) {
> 	return term(context.Background(), r, r.store.Engine(), i)
> ```

`term()` may make multiple calls into the supplied `engine.Reader`. If we don't make a snapshot then there is no guarantee of consistency. We need some sort of lazy snapshot `engine.Reader` that doesn't eagerly create the snapshot if we can hit the cache. 

---

*[storage/replica_raftstorage.go, line 397 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPEES1tnccn6CTKcdlV:-KPEES1tnccn6CTKcdlW:b-u73wke) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/replica_raftstorage.go#L397)):*
> ```Go
> func snapshot(
> 	ctx context.Context,
> 	r *Replica,
> ```

The use of a `RangeID` rather than a `Replica` here (and in all downstream functions - everything that's a static function instead of a method of `Replica`) was deliberate. This is run concurrently with other operations so we *must* get everything we need from `snap` instead of reaching into `Replica.store.Engine()`. The lower-level functions should take a `RangeID` and an optional `raftEntryCache` instead of a `Replica`; snapshots should never use the cache. 

---

*[storage/replica_test.go, line 5009 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPEAVSf5NMZWRh-HNfA:-KPEAVSgDdT3dEI1droU:bl157dx) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/replica_test.go#L5009)):*
> ```Go
> 	var indexes []uint64
> 
> 	populateLogs := func(from, to int) {
> ```

Return `indexes` from this function instead of mutating the outer variable. 

---

*[storage/replica_test.go, line 5041 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPEAy_eWex0aOELRMWw:-KPEAy_eWex0aOELRMWx:bcj47ax) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/replica_test.go#L5041)):*
> ```Go
> 		expError       error
> 	}{
> 		// Case 0: Just most of the entries from cache.
> ```

s/most/all/

---

*[storage/replica_test.go, line 5095 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPECXAtFJOPsbAY0DBU:-KPECXAtFJOPsbAY0DBV:b-q6eapj) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/replica_test.go#L5095)):*
> ```Go
> 	} {
> 		// Construct test data for test cases.
> 		if i == 4 {
> ```

This is pretty ugly; it's not clear in the test cases above that things are changing (persistently) between them. Maybe add a `func()` to the test case structure so the setup code can be inline with the cases? (if we can't make a single setup work for all cases)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494)*
<!-- Sent from Reviewable.io -->

petermattis commented:
@bdarnell We could also parallelize `Replica.tick`. Currently we `tick` every replica serially. 

---

Review status: all files reviewed at latest revision, 18 unresolved discussions, all commit checks successful.

---

*[storage/entry_cache.go, line 40 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPELAWHKzrJu9fzaMLM:-KPELAWHKzrJu9fzaMLN:b-4kdklo) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/entry_cache.go#L40)):*
> ```Go
> 	sync.Mutex                              // protects Cache for concurrent access.
> 	bytes        uint64                     // total size of the cache in bytes
> 	cache        *cache.UnorderedCache      // LRU cache of log entries, keyed by rangeID / log index
> ```

I'm mildly surprised this is not an `OrderedCache`. Seems like you want to find all of the entries from a given range in order, right? 

---

*[storage/entry_cache.go, line 52 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPEEmzKy0geYsr6o4aJ:-KPEEmzLjs2EsTgvtVq6:b9tumhq) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/entry_cache.go#L52)):*
> ```Go
> 	}
> 	// We expect the raft entry cache mutex to be held on ShouldEvict
> 	// and OnEvicted callbacks.
> ```

I'd rephrase this to: `The raft entry cache mutex will be held when the ShouldEvict and OnEvicted callbacks are invoked.`

PS Add a blank line between paragraphs.

---

*[storage/entry_cache.go, line 53 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPEEwy9m8gC72dQepw8:-KPEEwy9m8gC72dQepw9:b-oliotr) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/entry_cache.go#L53)):*
> ```Go
> 	// We expect the raft entry cache mutex to be held on ShouldEvict
> 	// and OnEvicted callbacks.
> 	// On ShouldEvice, compare the total size of the cache in bytes to the
> ```

`s/ShouldEvice/ShouldEvict/g`

---

*[storage/entry_cache.go, line 83 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPEIe4EvaRHt4Mu6tfr:-KPEIe4EvaRHt4Mu6tfs:bdh3z5o) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/entry_cache.go#L83)):*
> ```Go
> 
> 	if fi, ok := rec.firstIndexes[rangeID]; !ok || ents[0].Index < fi {
> 		rec.firstIndexes[rangeID] = ents[0].Index
> ```

Looks like this assumes `ents` is sorted. Perhaps worth checking in the loop over ents above.

How is `firstIndexes` kept up to date when an entry is evicted?

---

*[storage/entry_cache.go, line 87 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPEJP2VWGb6Dxe4HDmk:-KPEJP2W5t8mn4W1rE93:b-9t6rl1) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/entry_cache.go#L87)):*
> ```Go
> }
> 
> // getEntries returns entries between [lo, hi) for specified range.
> ```

Looks like this method gets a contiguous range of entries, stopping when the first missing entry is encountered. Let's spell this out in the comments and why it is important.

---

*[storage/entry_cache_test.go, line 61 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPEJZKeCj7UjMyqoBEY:-KPEJZKeCj7UjMyqoBEZ:b-bjfpw6) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/entry_cache_test.go#L61)):*
> ```Go
> 		t.Errorf("expected only two entries; got %+v, %d", ents, hi)
> 	}
> }
> ```

I'm not seeing a test which tests cache eviction. In particular, I'm interesting in how `firstIndexes` interacts with cache eviction.

---

*[storage/replica_raftstorage.go, line 200 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPEFCFogmqAEVo3i_zH:-KPEK90TU9QNSxvMhcmc:b3flim2) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/replica_raftstorage.go#L200)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> `term()` may make multiple calls into the supplied `engine.Reader`. If we don't make a snapshot then there is no guarantee of consistency. We need some sort of lazy snapshot `engine.Reader` that doesn't eagerly create the snapshot if we can hit the cache. 
</details>
RocksDB snapshots are lightweight. I'd revert this part of the change.

Alternately, we could change `term()` and the code below it to take an iterators. RocksDB iterators have an implicit snapshot associated with them (i.e. they don't see values written after the iterator was created). I think we have all of the infrastructure in the `MVCC*` routines to do this, but I think it best to leave to a separate PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-:-KPEI7tbD3Yv56zUk3-D:bu7irui)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 18 unresolved discussions, all commit checks successful.

---

*[storage/entry_cache.go, line 62 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPEJ6dvx_-Okv5urjSQ:-KPF0PqxNl0tpj-LJk8D:bjvr06f) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/entry_cache.go#L62)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> We need to update `firstIndex` on eviction. 
</details>
Going with an `OrderedCache` allowed removal of the `firstIndexes` map.

---

*[storage/entry_cache.go, line 83 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPEIe4EvaRHt4Mu6tfr:-KPF0WprzHNDrk0HxwJZ:blv90c5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/entry_cache.go#L83)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Looks like this assumes `ents` is sorted. Perhaps worth checking in the loop over ents above.
> 
> How is `firstIndexes` kept up to date when an entry is evicted?
</details>
Removed `firstIndexes`

---

*[storage/entry_cache.go, line 124 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPEILMeKKconR1tkWlT:-KPF1FNLMkKJN8SoK_8b:b-o9c7ek) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/entry_cache.go#L124)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> There's no guarantee that `hi` is in the cache (and with this data structure we have no way of finding the next remaining index). If there are multiple discontinuous cached spans, we'll have a `firstIndex` that is below the actual first index. This could make `clearTo` unexpectedly expensive as it has to iterate from an old `firstIndex` to the requested index. Maybe this should be an `OrderedCache` so we don't have to track the `firstIndexes` separate from the cache contents.
</details>
Changed to an `OrderedCache`.

---

*[storage/entry_cache_test.go, line 61 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPEJZKeCj7UjMyqoBEY:-KPFAQrsra58BiUF4lTf:bd9h38u) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/entry_cache_test.go#L61)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I'm not seeing a test which tests cache eviction. In particular, I'm interesting in how `firstIndexes` interacts with cache eviction.
</details>
`firstIndexes` is gone now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 2 of 10 files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---

*[storage/entry_cache.go, line 45 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPFm7Cgp2Jul8fnheyG:-KPFm7ChETcAUWNlAtqn:b-w0ps63) ([raw file](https://github.com/cockroachdb/cockroach/blob/012d07c38f5bfa2d6d62b851c0df2e8e90e2677d/storage/entry_cache.go#L45)):*
> ```Go
> 		return 1
> 	}
> 	return int(a.Index - bk.Index)
> ```

Use comparison here as this is buggy for certain input values: https://play.golang.org/p/zPG60uZ17D. I find this more palatable as a switch:

```
switch {
case a.RangeID < bk.RangeID:
  return -1
case a.RangeID > bk.RangID:
  return 1
case a.Index < bk.Index:
  return -1
case a.Index > bk.Index:
  return 1
default:
  return 0
}
```

---

*[storage/entry_cache.go, line 75 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPFnKmImOOdopEGIhaf:-KPFnKmImOOdopEGIhag:b-1bw3ik) ([raw file](https://github.com/cockroachdb/cockroach/blob/012d07c38f5bfa2d6d62b851c0df2e8e90e2677d/storage/entry_cache.go#L75)):*
> ```Go
> 	}
> 	rec.cache.Config.OnEvicted = func(k, v interface{}) {
> 		log.Infof(context.TODO(), "evicting log entry %+v", k)
> ```

Remove before merging?

---

*[storage/entry_cache.go, line 111 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPFnnbn7iruBFvHK3bb:-KPFnnbn7iruBFvHK3bc:buq8bay) ([raw file](https://github.com/cockroachdb/cockroach/blob/012d07c38f5bfa2d6d62b851c0df2e8e90e2677d/storage/entry_cache.go#L111)):*
> ```Go
> 	var bytes uint64
> 
> 	for i := lo; i < hi; i++ {
> ```

It would be faster to implement this using `OrderedCache.DoRange`. That iteration is faster than repeatedly looking up individual entries in the cache.

---

*[storage/entry_cache.go, line 151 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPFo311w5MIoGh1wBPm:-KPFo311w5MIoGh1wBPn:b-sxwjpw) ([raw file](https://github.com/cockroachdb/cockroach/blob/012d07c38f5bfa2d6d62b851c0df2e8e90e2677d/storage/entry_cache.go#L151)):*
> ```Go
> 		keys = append(keys, k.(entryCacheKey))
> 		return false
> 	}, entryCacheKey{RangeID: rangeID, Index: 0}, entryCacheKey{RangeID: rangeID, Index: index})
> ```

Hard to read the `DoRange` arguments like this. Might be worthwhile pulling out the second and third arguments into `from` and `to` variables to make them clearer.

---

*[storage/entry_cache_test.go, line 61 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPEJZKeCj7UjMyqoBEY:-KPFod_RwrgR0n7wjCR-:b-akj25a) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/entry_cache_test.go#L61)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> `firstIndexes` is gone now.
</details>
Ok. The testing here still seems a bit weak. Data structures like `RaftEntryCache` are so easy to test thoroughly in comparison to some of our other code. We should do so.

---

*[storage/store.go, line 528 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPFtwf_g8EnTFpFkXRG:-KPFtwf_g8EnTFpFkXRH:bcqgn5g) ([raw file](https://github.com/cockroachdb/cockroach/blob/012d07c38f5bfa2d6d62b851c0df2e8e90e2677d/storage/store.go#L528)):*
> ```Go
> 	// RaftEntryCacheSize is the size in bytes of the Raft log entry cache
> 	// shared by all Raft groups managed by the store.
> 	RaftEntryCacheSize uint64
> ```

We don't ever seem to set this field.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 2 of 10 files reviewed at latest revision, 22 unresolved discussions, some commit checks failed.

---

*[storage/replica_raftstorage.go, line 200 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPEFCFogmqAEVo3i_zH:-KPIZBCUpsGtGcuZzmGY:b2iq15c) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/replica_raftstorage.go#L200)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> RocksDB snapshots are lightweight. I'd revert this part of the change.
> 
> Alternately, we could change `term()` and the code below it to take an iterators. RocksDB iterators have an implicit snapshot associated with them (i.e. they don't see values written after the iterator was created). I think we have all of the infrastructure in the `MVCC*` routines to do this, but I think it best to leave to a separate PR.
</details>
Thinking about this again, aren't we protected by `Store.processRaftMu` here? There shouldn't be any concurrent writes to the replica's on-disk state going on. I'd prefer to leave this as-is for this PR, but Spencer's change may have been safe.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 2 of 10 files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---

*[kv/range_cache.go, line 630 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPIlcTx2VNeZaLn7UYf:-KPIlcTx2VNeZaLn7UYg:b-pzlvl4) ([raw file](https://github.com/cockroachdb/cockroach/blob/012d07c38f5bfa2d6d62b851c0df2e8e90e2677d/kv/range_cache.go#L630)):*
> ```Go
> 	// TODO(spencer): uh oh, deleting while within DoRange looks like it
> 	// doesn't do the right thing. Need to change this to accumulate keys
> 	// to delete and delete those in pass 2.
> ```

I already filed an issue (#8564) about this, no need for this TODO.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 2 of 10 files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---

*[storage/replica_raftstorage.go, line 200 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPEFCFogmqAEVo3i_zH:-KPIss7Rhy9tSyiiiMJ4:bonmpl8) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/replica_raftstorage.go#L200)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Thinking about this again, aren't we protected by `Store.processRaftMu` here? There shouldn't be any concurrent writes to the replica's on-disk state going on. I'd prefer to leave this as-is for this PR, but Spencer's change may have been safe.
</details>
Maybe, but it's subtle, and I don't think it's guaranteed by our current concurrency documentation. `Term()` is documented to require `Replica.mu`, but we don't hold `Replica.mu` while writing to rocksdb. I think it's currently the case that we hold `Store.processRaftMu` whenever we either call `Term()` or write to rocksdb, but I'd rather avoid relying on this much subtlety in our concurrency.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 2 of 10 files reviewed at latest revision, 24 unresolved discussions, some commit checks failed.

---

*[storage/entry_cache_test.go, line 61 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPEJZKeCj7UjMyqoBEY:-KPJwPb5FchUtFGJWRGg:b8zcp0z) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/entry_cache_test.go#L61)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Ok. The testing here still seems a bit weak. Data structures like `RaftEntryCache` are so easy to test thoroughly in comparison to some of our other code. We should do so.
</details>
Anything specific you're after?

---

*[storage/replica_raftstorage.go, line 200 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPEFCFogmqAEVo3i_zH:-KPJvrnx4qtvPxBQQoq1:b-ex2oye) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/replica_raftstorage.go#L200)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Maybe, but it's subtle, and I don't think it's guaranteed by our current concurrency documentation. `Term()` is documented to require `Replica.mu`, but we don't hold `Replica.mu` while writing to rocksdb. I think it's currently the case that we hold `Store.processRaftMu` whenever we either call `Term()` or write to rocksdb, but I'd rather avoid relying on this much subtlety in our concurrency.
</details>
Will leave with snapshot here.

---

*[storage/store.go, line 528 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPFtwf_g8EnTFpFkXRG:-KPJvv-7mAoqRfvsV1vV:b-9k3ec) ([raw file](https://github.com/cockroachdb/cockroach/blob/012d07c38f5bfa2d6d62b851c0df2e8e90e2677d/storage/store.go#L528)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> We don't ever seem to set this field.
</details>
It gets the default value right now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 2 of 9 files reviewed at latest revision, 26 unresolved discussions, some commit checks pending.

---

*[storage/entry_cache.go, line 142 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPKE_w2mptd0ryZjis8:-KPKE_w2mptd0ryZjis9:b-nyt55b) ([raw file](https://github.com/cockroachdb/cockroach/blob/915db0ce8a6f1036c200165e9348703e3b431bf5/storage/entry_cache.go#L142)):*
> ```Go
> 	}
> 
> 	for i := lo; i < hi; i++ {
> ```

This is inefficient if `lo` is much smaller than `hi`. Might be worth searching for the first cached entry and then iterating forward from there.

Is this only used in tests? If yes, disregard this suggestion and perhaps move into the `storage_test` package.

---

*[storage/entry_cache_test.go, line 61 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPEJZKeCj7UjMyqoBEY:-KPKDCqtnz8wJAFMUGh3:bsmsqjb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f6cf356fa5d00f1991669689899963bce453a76/storage/entry_cache_test.go#L61)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Anything specific you're after?
</details>
I'd like to see testing of the various cases for the data structure:
* `getEntries` where the cached data lies before the range we're retrieving
* `getEntries` where the cached data lies after the range we're retrieving
* `getEntries` where the cached data overlaps the range we're retrieving
* adding entries with gaps and testing the behavior of `getEntries`
*  various `delEntries` conditions:
  - deleting a prefix of the cached entries
  - deleting a suffix of the cached entries
  - creating a gap in the middle of the cached entries

---

*[storage/replica_raftstorage.go, line 128 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPKFUUbIxfh-0vBcUc9:-KPKFUUbIxfh-0vBcUcA:b-goiuaj) ([raw file](https://github.com/cockroachdb/cockroach/blob/915db0ce8a6f1036c200165e9348703e3b431bf5/storage/replica_raftstorage.go#L128)):*
> ```Go
> 	}
> 
> 	if err := iterateEntries(ctx, e, rangeID, newLo, hi, scanFunc); err != nil {
> ```

Can you `s/newLo/expectedIndex/g` and get rid of `newLo`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-:-KPKG5mJSENSd6AVdALM:bnfp4nl)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 2 of 9 files reviewed at latest revision, 26 unresolved discussions.

---

*[storage/entry_cache.go, line 142 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-KPKE_w2mptd0ryZjis8:-KPNrVOCtYjZJwQSFLk2:blrpaf8) ([raw file](https://github.com/cockroachdb/cockroach/blob/915db0ce8a6f1036c200165e9348703e3b431bf5/storage/entry_cache.go#L142)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> This is inefficient if `lo` is much smaller than `hi`. Might be worth searching for the first cached entry and then iterating forward from there.
> 
> Is this only used in tests? If yes, disregard this suggestion and perhaps move into the `storage_test` package.
</details>
I changed `clearTo` to call `delEntries` and moved code accordingly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 8 files at r2, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 26 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8494#-:-KPOAlRuSuGoprZ17204:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Should this be targeted to develop rather than master?

bdarnell commented:
I want this on master because I think it will help with some of the issues we're seeing on beta.
